moveit_core/trajectory_processing/src/cartesian_speed.cpp

```diff
@@ -0,0 +1,106 @@
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Benjamin Scholz */
+
+#include <moveit/trajectory_processing/cartesian_speed.h>
+
+// Name of logger
+const char* LOGGER_NAME = "trajectory_processing.cartesian_speed";
+
+namespace trajectory_processing
+{
+bool setMaxCartesianEndEffectorSpeed(robot_trajectory::RobotTrajectory& trajectory, const double max_speed,
+                                     std::string end_effector)
+{
+  if (max_speed <= 0.0)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGGER_NAME, "End effector speed must be greater than 0.");
+    return false;
+  }
+  if (end_effector.empty())
+  {
+    std::vector<std::string> tips;
+    trajectory.getGroup()->getEndEffectorTips(tips);
+    if (tips.empty())
+    {
+      ROS_ERROR_STREAM_NAMED(LOGGER_NAME, "No end effector defined for group attached to trajectory.");
+      return false;
+    }
+    end_effector = tips[0];
+  }
+
+  size_t num_waypoints = trajectory.getWayPointCount();
+  if (num_waypoints == 0)
+    return false;
+
+  robot_state::RobotStatePtr kinematic_state = trajectory.getFirstWayPointPtr();
+
+  // do forward kinematics to get cartesian positions of end effector for current waypoint
+  double euclidean_distance, new_time_diff, old_time_diff;
+  std::vector<double> time_diff(num_waypoints - 1, 0.0);
+
+  for (size_t i = 0; i < num_waypoints - 1; i++)
+  {
+    // get end effector state for current waypoint
+    Eigen::Isometry3d current_end_effector_state = trajectory.getWayPointPtr(i)->getGlobalLinkTransform(end_effector);
+
+    // get end effector state for next waypoint
+    Eigen::Isometry3d next_end_effector_state = trajectory.getWayPointPtr(i + 1)->getGlobalLinkTransform(end_effector);
+
+    // get euclidean distance between the two waypoints
+    euclidean_distance = (next_end_effector_state.translation() - current_end_effector_state.translation()).norm();
+
+    new_time_diff = (euclidean_distance / max_speed);
+    old_time_diff = trajectory.getWayPointDurationFromPrevious(i + 1);
+
+    // if constraints allow save the new time difference between waypoints
+    if (new_time_diff > old_time_diff)
+    {
+      time_diff[i] = new_time_diff;
+    }
+    else
+    {
+      ROS_WARN_ONCE_NAMED(LOGGER_NAME,
+                          "Desired cartesian end-effector speed is not reached because of joint velocity constraints.");
```

Comment:
I'm not happy with this being a `ONCE` warning in a multi-purpose method.
Either this is a critical failure condition for the method (in which case it should not be called `Max`), or it is perfectly fine behavior.
In this case, we could report it to the user by returning a `double` value that represents the **slowest** Cartesian speed in the adapted trajectory.

Reply:
I guess you resolved this because you [added the slowest speed into the ROS_WARN](https://github.com/ros-planning/moveit/pull/2674/files#diff-977a18c3ac8d4467d1fb062651186b55476f7ecb589a3ec2f3c665c6f6c9509dR128), but that's not what I meant.
The method is called `*limitMax*CartesianLinkSpeed` and by name uses would not expect a warning here at all!
The only solutions I see here are those I outlined above:

(1) Either remove the whole warning and accept it as normal method semantics, or 
(2) rename the method to `enforceCartesianLinkSpeed` or similar and make it a critical error in this method

I would prefer (1) by far, also because acceleration periods (esp. for high-density trajectories) **always** limit the Cartesian speed if the start/end waypoints have 0 velocities.